pinot-core/src/main/java/org/apache/pinot/core/common/datablock/DataBlockBuilder.java

```diff
@@ -187,200 +147,321 @@ public static RowDataBlock buildFromRows(List<Object[]> rows, DataSchema dataSch
-      columnarBuilder.setNullRowIds(nullBitmap);
+    return rowSizeInBytes;
+  }
+
+  private static void writeVarOffsetInFixed(ByteBuffer fixedSize, PagedPinotOutputStream varSize) {
+    long offsetInVar = varSize.getCurrentOffset();
+    Preconditions.checkState(offsetInVar <= Integer.MAX_VALUE,
+        "Cannot handle variable size output stream larger than 2GB");
+    fixedSize.putInt((int) offsetInVar);
+  }
+
+  private static void setNullRowIds(RoaringBitmap[] nullVectors, ByteBuffer fixedSize,
+      CompoundDataBuffer.Builder varBufferBuilder)
+      throws IOException {
+    int varBufSize = Arrays.stream(nullVectors)
+        .mapToInt(bitmap -> bitmap == null ? 0 : bitmap.serializedSizeInBytes())
+        .sum();
+    ByteBuffer variableSize = ByteBuffer.allocate(varBufSize)
+        .order(ByteOrder.BIG_ENDIAN);
+
+    long varWrittenBytes = varBufferBuilder.getWrittenBytes();
+    Preconditions.checkArgument(varWrittenBytes < Integer.MAX_VALUE,
+        "Cannot handle variable size output stream larger than 2GB but found {} written bytes", varWrittenBytes);
+    int startVariableOffset = (int) varWrittenBytes;
+    for (RoaringBitmap nullRowIds : nullVectors) {
+      int writtenVarBytes = variableSize.position();
+      fixedSize.putInt(startVariableOffset + writtenVarBytes);
+      if (nullRowIds == null || nullRowIds.isEmpty()) {
+        fixedSize.putInt(0);
+      } else {
+        RoaringBitmapUtils.serialize(nullRowIds, variableSize);
+        fixedSize.putInt(variableSize.position() - writtenVarBytes);
+      }
     }
-    return buildColumnarBlock(columnarBuilder);
+    varBufferBuilder.addBuffer(variableSize);
   }
 
-  private static RowDataBlock buildRowBlock(DataBlockBuilder builder) {
-    return new RowDataBlock(builder._numRows, builder._dataSchema, getReverseDictionary(builder._dictionary),
-        builder._fixedSizeDataByteArrayOutputStream.toByteArray(),
-        builder._variableSizeDataByteArrayOutputStream.toByteArray());
+  private static RowDataBlock buildRowBlock(int numRows, DataSchema dataSchema, String[] dictionary,
+      ByteBuffer fixedSize, CompoundDataBuffer.Builder varBufferBuilder) {
+    return new RowDataBlock(numRows, dataSchema, dictionary, PinotByteBuffer.wrap(fixedSize), varBufferBuilder.build());
   }
 
-  private static ColumnarDataBlock buildColumnarBlock(DataBlockBuilder builder) {
-    return new ColumnarDataBlock(builder._numRows, builder._dataSchema, getReverseDictionary(builder._dictionary),
-        builder._fixedSizeDataByteArrayOutputStream.toByteArray(),
-        builder._variableSizeDataByteArrayOutputStream.toByteArray());
+  private static ColumnarDataBlock buildColumnarBlock(int numRows, DataSchema dataSchema, String[] dictionary,
+      ByteBuffer fixedSize, CompoundDataBuffer.Builder varBufferBuilder) {
+    return new ColumnarDataBlock(numRows, dataSchema, dictionary,
+        PinotByteBuffer.wrap(fixedSize), varBufferBuilder.build());
```

Comment:
nit: These can probably be simply in-lined now?

Reply:
What do you mean? Just to write it like

```
return new ColumnarDataBlock(numRows, dataSchema, dictionary, PinotByteBuffer.wrap(fixedSize), varBufferBuilder.build());
```

In that case we cannot do that because that line is 126 chars long